bindings/bind_py_structure_manager.cc

```diff
@@ -616,6 +618,196 @@ namespace rascal {
         R"(Get the dense feature matrix associated with the calculator and
         the collection of structures (managers) using the list of keys
         provided. Only applicable when Calculator uses BlockSparseProperty.)");
+    manager_collection.def(
+        "get_features_gradient",
+        [](ManagerCollection_t & managers, const Calculator & calculator,
+           py::list & all_keys_l) {
+          using Manager_t = typename ManagerCollection_t::Manager_t;
+          using Prop_t =
+              typename Calculator::template PropertyGradient_t<Manager_t>;
+          using Keys_t = typename Prop_t::Keys_t;
+          if (managers.size() == 0) {
+            throw std::runtime_error(
+                R"(There are no structure to get features from)");
+          }
+          Keys_t all_keys;
+          if (all_keys_l.size() > 0) {
+            // convert the list of keys from python in to the proper type
+            for (py::handle key_l : all_keys_l) {
+              auto key = py::cast<std::vector<int>>(key_l);
+              all_keys.insert(key);
+            }
+          } else {
+            // empty all_keys_l means that we use the keys from the managers
+            for (auto key_l : managers.get_keys(calculator)) {
+              all_keys.insert(key_l);
+            }
+          }
+
+          auto property_name{managers.get_calculator_name(calculator, true)};
+
+          const auto & property_ =
+              *managers[0]->template get_property<Prop_t>(property_name);
+          // assume inner_size is consistent for all managers
```

Comment:
is there a way to check this assumption? Will the copy in the full matrix below will fail with a runtime error in this case?

Reply:
the assumption is valid since it's the same `property_name` which by default encodes a uniform `inner_size`. Nevertheless I guess that some twisted user could in principle have non uniform `inner_size` in a set of structures but he would have had to try hard to get to this point...